src/vm/eventpipe.cpp

```diff
@@ -343,34 +335,34 @@ void EventPipe::Disable(EventPipeSessionID id)
 
     SetupThread();
 
-    // Only perform the disable operation if the session ID
-    // matches the current active session.
-    if (id != (EventPipeSessionID)s_pSession)
+    if (id == 0 || !Enabled())
         return;
 
     // Don't block GC during clean-up.
     GCX_PREEMP();
 
-    EventPipe::RunWithCallbackPostponed(
-        [&](EventPipeProviderCallbackDataQueue* pEventPipeProviderCallbackDataQueue)
-        {
-            DisableInternal(reinterpret_cast<EventPipeSessionID>(s_pSession), pEventPipeProviderCallbackDataQueue);
-        }
-    );
+    RunWithCallbackPostponed([&](EventPipeProviderCallbackDataQueue *pEventPipeProviderCallbackDataQueue) {
+        EventPipeSession *pSession = reinterpret_cast<EventPipeSession *>(id);
+        if (!s_pSessions->Contains(pSession))
+            return;
+        DisableInternal(*pSession, pEventPipeProviderCallbackDataQueue);
+        s_pSessions->Remove(pSession);
+        delete pSession;
+    });
 }
 
-void EventPipe::DisableInternal(EventPipeSessionID id, EventPipeProviderCallbackDataQueue* pEventPipeProviderCallbackDataQueue)
+void EventPipe::DisableInternal(EventPipeSession &session, EventPipeProviderCallbackDataQueue* pEventPipeProviderCallbackDataQueue)
 {
     CONTRACTL
     {
         THROWS;
         GC_TRIGGERS;
         MODE_ANY;
-        PRECONDITION(GetLock()->OwnedByCurrentThread());
+        PRECONDITION(IsLockOwnedByCurrentThread());
```

Comment:
I'm fine if you want to do it in a later PR, but something should probably track breaking the lock when doing Disable(), otherwise Disable() on session A will be able to block all other sessions from doing work for a long time.

Reply:
I'll postpone it to future PR.